org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/commands/BuildPathCommand.java

```diff
@@ -0,0 +1,242 @@
+		IProject[] projects = ProjectUtils.getAllProjects();
+		for (IProject project : projects) {
+			if (!ProjectsManager.DEFAULT_PROJECT_NAME.equals(project.getName()) && ProjectUtils.isJavaProject(project)) {
+				try {
+					IPath[] paths = ProjectUtils.listSourcePaths(JavaCore.create(project));
+					for (IPath path : paths) {
+						IPath entryPath = path;
+						String projectName = project.getName();
+						String projectType = "General";
+						if (ProjectUtils.isMavenProject(project)) {
+							projectType = "Maven";
+						}
+
+						if (ProjectUtils.isGradleProject(project)) {
+							projectType = "Gradle";
+						}
+
+						IContainer projectRoot = project;
+						if (!ProjectUtils.isVisibleProject(project)) {
+							projectType = "Workspace";
+							IFolder workspaceLinkFolder = project.getFolder(ProjectUtils.WORKSPACE_LINK);
+							if (!workspaceLinkFolder.isLinked()) {
+								continue;
+							}
+
+							projectRoot = workspaceLinkFolder;
+						}
+
+						IPath relativePath = entryPath.makeRelativeTo(projectRoot.getFullPath());
+						IPath location = projectRoot.getRawLocation().append(relativePath);
+						IPath displayPath = getWorkspacePath(location);
+						sourcePathList.add(new SourcePath(location != null ? location.toOSString() : "", displayPath != null ? displayPath.toOSString() : entryPath.toOSString(), projectName, projectType));
+					}
+				} catch (JavaModelException e) {
+					JavaLanguageServerPlugin.logException("Failed to resolve the existing source paths in current workspace.", e);
+					return new CommandResult(false, e.getMessage());
+				}
+			}
+		}
+
+		return new CommandResult(true, null, sourcePathList.toArray(new SourcePath[0]));
+	}
+
+	private static IProject findBelongedProject(IPath sourceFolder) {
+		List<IProject> projects = Stream.of(ProjectUtils.getAllProjects()).filter(ProjectUtils::isJavaProject).sorted(new Comparator<IProject>() {
+			@Override
+			public int compare(IProject p1, IProject p2) {
+				return p2.getLocation().toOSString().length() - p1.getLocation().toOSString().length();
+			}
+		}).collect(Collectors.toList());
+
+		for (IProject project : projects) {
+			if (project.getLocation().isPrefixOf(sourceFolder)) {
+				return project;
+			}
+		}
+
+		return null;
+	}
+
+	private static IPath getWorkspacePath(IPath path) {
+		PreferenceManager manager = JavaLanguageServerPlugin.getPreferencesManager();
+		Collection<IPath> rootPaths = manager.getPreferences().getRootPaths();
+		if (rootPaths != null) {
+			for (IPath rootPath : rootPaths) {
+				if (rootPath.isPrefixOf(path)) {
+					return path.makeRelativeTo(rootPath.append(".."));
+				}
+			}
+		}
+
+		return path;
+	}
+
+	public static class CommandResult {
+		public boolean status;
+		public String message;
+		public Object data;
+
+		CommandResult(boolean status, String message) {
+			this.status = status;
+			this.message = message;
+		}
+
+		CommandResult(boolean status, String message, Object data) {
```

Comment:
If the result holds data of different types, we should have different result classes.